### PR TITLE
2170: Warn on trailing period in PR titles

### PR DESCRIPTION
**Pull request by zhaosongzs**
This patch is trying to add a new jcheck called "issuestitle". This jcheck would check if the issue's title has a trailing period or leading lowercase letter. And we would like to configure "issuestitle" jcheck as warning in the repos since in some cases, trailing periods or leading lowercase letter are valid.

When implementing this, I found that although we can configure a jcheck as warning in the jcheck configuration file(.jcheck/conf), we fail to differentiate between jchecks as error and jchecks as warning. When jchecks fail as warnings, they cause the local jcheck to exit with a status code of 1. Also, they will be integration blockers when Skara bots evaluate pull requests.  Therefore, in this patch, I also make Skara CLI and Skara bots be able to handle jcheck warnings properly.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issues
 * [SKARA-2170](https://bugs.openjdk.org/browse/SKARA-2170): Warn on trailing period in PR titles (**Enhancement** - P4)
 * [SKARA-2248](https://bugs.openjdk.org/browse/SKARA-2248): Warn on leading lowercase letter in PR titles (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1638/head:pull/1638` \
`$ git checkout pull/1638`

Update a local copy of the PR: \
`$ git checkout pull/1638` \
`$ git pull https://git.openjdk.org/skara.git pull/1638/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1638`

View PR using the GUI difftool: \
`$ git pr show -t 1638`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1638.diff">https://git.openjdk.org/skara/pull/1638.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1638#issuecomment-2061996795)

### CODE REVIEW

**Comment by magicus** [comment]
I looked at the code to try to answer my own question, but could not find out an easy answer, so I'll say it out loud instead:

How do you make it so that this is a warning only, and not a blocking error?

**Comment by zhaosongzs** [comment]
> I looked at the code to try to answer my own question, but could not find out an easy answer, so I'll say it out loud instead:
> 
> How do you make it so that this is a warning only, and not a blocking error?

Hi Magnus, I haven't finished this patch... So it's still in the draft state. I will finish it today.

**Comment by zhaosongzs** [comment]
ping

**Comment by zhaosongzs** [comment]
This PR is ready for review. Could someone please review it? ping...

**Comment by erikj79** [comment]
> This PR is ready for review. Could someone please review it? ping...

Sorry, I saw your comment about it not being ready so assumed you would say something when it was. Noticed the draft state change now.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java [hunk 1]
```diff
@@ -54,16 +54,27 @@ class PullRequestCheckIssueVisitor implements IssueVisitor {
...
     }
 
-    private void addFailureMessage(Check check, String message) {
-        failedChecks.put(check.getClass(), message);
+    private void addMessage(Check check, String message, Severity severity, boolean readyForReviewWhenFailedAsError) {
+        if (severity.equals(Severity.ERROR)) {
+            errorFailedChecks.put(check.getClass(), message);
+            if (this.readyForReview) {
+                this.readyForReview = readyForReviewWhenFailedAsError;
+            }
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java:62]: This part is rather confusing to try to understand. Maybe we need a separate method for updating the `readyForReview` field, and remove the new parameter `readyForReviewWhenFailedAsError` from this method, to make this interaction clearer? 

```
private void setNotReadyForReviewOnError(Severity severity) {
    if (severity == ERROR) {
        readyForReview = false;
    }
}
```

This method would be called from the places that used to do `readyForReview = false`.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java [hunk 2]
```diff
@@ -252,37 +259,45 @@ public void visit(MessageWhitespaceIssue issue) {
...
     public void visit(ProblemListsIssue issue) {
-        addFailureMessage(issue.check(), issue.issue() + " is used in problem lists: " + issue.files());
+        addMessage(issue.check(), issue.issue() + " is used in problem lists: " + issue.files(), issue.severity(), true);
+    }
+
+    @Override
+    public void visit(IssuesTitleIssue issue) {
+        List<String> messages = new ArrayList<>();
+        if (!issue.issuesWithTrailingPeriod().isEmpty()) {
+            messages.add("Found trailing period in " + String.join(" ,", issue.issuesWithTrailingPeriod()));
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java:295]: Shouldn't the join be comma followed with space? Maybe specify that this is in the issue title?
```suggestion
            messages.add("Found trailing period in issue title for " + String.join(", ", issue.issuesWithTrailingPeriod()));
```

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java [hunk 3]
```diff
@@ -252,37 +259,45 @@ public void visit(MessageWhitespaceIssue issue) {
...
+        }
+        if (!issue.issuesWithLeadingLowerCaseLetter().isEmpty()) {
+            messages.add("Found leading lowercase letter in " + String.join(" ,", issue.issuesWithLeadingLowerCaseLetter()));
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java:298]:
```suggestion
            messages.add("Found leading lowercase letter in issue title for " + String.join(" ,", issue.issuesWithLeadingLowerCaseLetter()));
```

File: cli/src/main/java/org/openjdk/skara/cli/JCheckCLIVisitor.java [hunk 4]
```diff
@@ -248,65 +248,81 @@ public void visit(IssuesIssue i) {
...
+
+    @Override
+    public void visit(IssuesTitleIssue i) {
+        if (!ignore.contains(i.check().name()) && !isLax) {
+            if (!i.issuesWithTrailingPeriod().isEmpty()) {
+                println(i, "Found trailing period in " + String.join(" ,", i.issuesWithTrailingPeriod()));
+            }
+            if (!i.issuesWithLeadingLowerCaseLetter().isEmpty()) {
+                println(i, "Found leading lowercase letter in " + String.join(" ,", i.issuesWithLeadingLowerCaseLetter()));
+            }
```
Comment (erikj79) [inline cli/src/main/java/org/openjdk/skara/cli/JCheckCLIVisitor.java:321]: Same comment here about wording and joining as above.

File: jcheck/src/main/java/org/openjdk/skara/jcheck/IssuesTitleCheck.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public String name() {
+        return "issuestitle";
+    }
+
+    @Override
+    public String description() {
+        return "Issue's title shouldn't have trailing period";
```
Comment (erikj79) [inline jcheck/src/main/java/org/openjdk/skara/jcheck/IssuesTitleCheck.java:71]:
```suggestion
        return "Issue's title should be properly formatted";
```

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thanks for the review!
/integrate

**Pull request closed without merging**